File: Tutorial/Py/TutorialNlpByteGruAg.py
#!/usr/bin/env python3
"""OA Python Tutorial — Byte-Level GRU, all-position LM (Autograd).

Python port of Tutorial/Ml/Nlp/TutorialNlpByteGruAg.cpp. The gated-recurrent member of
the NLP fair-comparison suite: same all-position
next-token task / corpus / dims as the RNN, swapping OaRnn -> OaGru (reset/update
gates, fused whole-sequence scan).

Model:     OaByteEmbedding -> OaGru -> OaLinear, projected at every step
Optimizer: AdamW ; Backward: implicit autograd via OaGradientTape

Run:
    OA_PYTHON_BUILD_DIR=Build/Release PYTHONPATH=Tutorial/Py \
      Build/PythonVenv/bin/python Tutorial/Py/TutorialNlpByteGruAg.py
"""

from __future__ import annotations

import math
import os
import tempfile

from _oa_import import core, ml, oa, runtime
import _nlp_common as nlp

VOCAB = nlp.BYTE_VOCAB_SIZE  # 256 — byte vocab family


class ByteGruLM:
    """OaByteEmbedding -> OaGru -> OaLinear, projecting every timestep to the vocab."""

    def __init__(self):
        wd = core.GetWeightDtype()

        self.embed = ml.OaByteEmbedding(nlp.D_MODEL)
        self.embed.Parameters()[0].Data = core.RandN(VOCAB, nlp.D_MODEL, wd)

        self.gru = ml.OaGru(nlp.D_MODEL, nlp.HIDDEN_DIM, 1)

        self.head = ml.OaLinear(nlp.HIDDEN_DIM, VOCAB)
        self.head.Parameters()[0].Data = core.Rand(VOCAB, nlp.HIDDEN_DIM, wd)

        for p in self.parameters():
            p.Data.SetRequiresGrad(True)

    def forward(self, x):
        b, s = x.Size(0), x.Size(1)
        e = self.embed.Forward(x).Reshape([b, s, nlp.D_MODEL])
        o = self.gru.Forward(e).Reshape([b * s, nlp.HIDDEN_DIM])
        return self.head.Forward(o)

    def parameters(self):
        return (list(self.embed.Parameters())
                + list(self.gru.AllParameterPtrs())
                + list(self.head.Parameters()))

    def save(self, prefix):
        self.embed.Save(prefix + ".embed.oam")
        self.gru.Save(prefix + ".gru.oam")
        self.head.Save(prefix + ".head.oam")

    def load(self, prefix):
        self.embed.Load(prefix + ".embed.oam")
        self.gru.Load(prefix + ".gru.oam")
        self.head.Load(prefix + ".head.oam")


def main():
    runtime.OaInitComputeEngine()
    print("\n" + "=" * 66)
    print("  OA Python Tutorial — Byte GRU · all-position LM (Autograd)")
    print("=" * 66)
    print(f"Vocab: {VOCAB} bytes · Context: {nlp.CONTEXT_LEN} · "
          f"DModel: {nlp.D_MODEL} · Hidden: {nlp.HIDDEN_DIM}")

    model = ByteGruLM()
    params = model.parameters()
    n_params = sum(p.Data.NumElements() for p in params)
    print(f"Model: ByteEmbed({VOCAB}→{nlp.D_MODEL}) → GRU({nlp.D_MODEL}→{nlp.HIDDEN_DIM}) "
          f"→ Linear({nlp.HIDDEN_DIM}→{VOCAB})")
    print(f"Params: {n_params}    Optimizer: AdamW(lr=0.01)\n")

    sampler = nlp.NlpAllPositionSampler(nlp.CORPUS, nlp.BATCH)
    opt, initial_loss, last_loss, x, y = nlp.train_all_position(
        model.forward, params, sampler, lr=0.01, timer_name="byte_gru_allpos_step")

    acc = nlp.accuracy_all_positions(model.forward, x, y, VOCAB)
    gen = nlp.generate_greedy(model.forward, "hello", 48, VOCAB)
    print(f"\nBaseline: ln({VOCAB}) = {math.log(VOCAB):.4f}")
    print(f"Initial loss: {initial_loss:.4f} · Final loss: {last_loss:.4f} · Accuracy: {acc:.1f}%")
    print(f"Generation:\n  Prompt:    'hello'\n  Generated: '{gen}'")

    prefix = os.path.join(tempfile.gettempdir(), "oa_py_byte_gru_allpos")
    model.save(prefix)
    reloaded = ByteGruLM()
    reloaded.load(prefix)
    reloaded_acc = nlp.accuracy_all_positions(reloaded.forward, x, y, VOCAB)
    print(f"\nSave/Load round-trip: reloaded accuracy {reloaded_acc:.1f}% "
          f"(trained {acc:.1f}%)")

    assert initial_loss > 0.0
    assert last_loss < initial_loss, "Loss must decrease during training"
    assert acc > 50.0, f"All-position accuracy should exceed 50%, got {acc:.1f}%"
    assert abs(reloaded_acc - acc) < 0.5, "Reloaded model must match trained accuracy"
    print("\n✓ All checks passed")


if __name__ == "__main__":
    main()

File: Tutorial/Ml/Nlp/TutorialNlpByteGruAg.cpp
// ═══════════════════════════════════════════════════════════════════════════
// OA Tutorial — Byte-Level GRU, all-position LM (Implicit Autograd)
// Module API: OaByteEmbedding + OaGru + OaLinear + OaAdamW + OaGradientTape
// ═══════════════════════════════════════════════════════════════════════════
//
// The GRU member of the NLP fair-comparison suite (TutorialNlpCommon.h). Same
// all-position next-token task, corpus and dims as the RNN/Transformer/Mamba-3/
// Empyrealm siblings — the gated recurrent unit adds reset/update gates over the
// plain RNN. We project every timestep to the vocab (a real LM), not just the last.
// ═══════════════════════════════════════════════════════════════════════════

#include "OaTest.h"
#include "TutorialMl.h"
#include "TutorialNlpCommon.h"
#include <Oa/Ml.h>
#include <Oa/Ml/Byte.h>
#include <Oa/Ml/Autograd.h>
#include <Oa/Runtime/Context.h>
#include <Oa/Runtime/Engine.h>
#include <cmath>

static constexpr OaI32 kVocabSize = OA_BYTE_VOCAB_SIZE;  // 256 — byte vocab family

// ─── Model: ByteEmbedding → GRU → Linear, projected at every step ────────────

class OaByteGruLM : public OaModule {
public:
	OaByteGruLM() {
		auto wd = OaFnMatrix::GetWeightDtype();

		Embed_ = OaMakeSharedPtr<OaByteEmbedding>(kDModel);
		Embed_->Parameters()[0].Data = OaFnMatrix::RandN(OaMatrixShape{kVocabSize, kDModel}, wd);
		Embed_->Parameters()[0].Data.SetRequiresGrad(true);
	
		Gru_ = OaMakeSharedPtr<OaGru>(kDModel, kHiddenDim, 1);
		for (auto& param : Gru_->Parameters()) {
			param.Data.SetRequiresGrad(true);
		}
	
		Head_ = OaMakeSharedPtr<OaLinear>(kHiddenDim, kVocabSize);
		Head_->Parameters()[0].Data = OaFnMatrix::Rand(OaMatrixShape{kVocabSize, kHiddenDim}, wd);
		for (auto& param : Head_->Parameters()) {
			param.Data.SetRequiresGrad(true);
		}

		RegisterModule("embed", Embed_);
		RegisterModule("gru",   Gru_);
		RegisterModule("head",  Head_);
	}

	OaMatrix Forward(const OaMatrix& InTokens) override {
		const OaI32 batch  = static_cast<OaI32>(InTokens.Size(0));
		const OaI32 seqLen = static_cast<OaI32>(InTokens.Size(1));

		// Gather/ByteEmbedding flattens to [batch*seq, d_model]; restore the rank-3
		// shape OaGru::Forward expects (else it slices the wrong axis → O(seq^2)).
		auto embedded = Embed_->Forward(InTokens).Reshape(OaMatrixShape{batch, seqLen, kDModel});
		auto gruOut   = Gru_->Forward(embedded);                          // [B, S, H]
		auto flat     = gruOut.Reshape(OaMatrixShape{static_cast<OaI64>(batch) * seqLen, kHiddenDim});  // [B*S, H]
		return Head_->Forward(flat);                                      // [B*S, V]
	}

private:
	OaSharedPtr<OaByteEmbedding> Embed_;
	OaSharedPtr<OaGru>           Gru_;
	OaSharedPtr<OaLinear>        Head_;
};

// ─── Tutorial ──────────────────────────────────────────────────────────────

TEST(TutorialNlpByteGruAg, GruAllPositionLM) {
	printf("\n╔══════════════════════════════════════════════════════════════════╗\n");
	printf("║  OA Tutorial — Byte GRU · all-position LM (Autograd)            ║\n");
	printf("╚══════════════════════════════════════════════════════════════════╝\n\n");
	printf("Vocab: %d bytes · Context: %d · DModel: %d · Hidden: %d\n",
		kVocabSize, kContextLen, kDModel, kHiddenDim);
	printf("Task: dense next-byte at every position via a gated recurrent unit\n\n");

	OaFnMatrix::SetRngSeed(OaNlpSuiteRngSeed);
	auto  model  = OaMakeSharedPtr<OaByteGruLM>();
	auto  params = model->AllParameterPtrs();
	auto  opt    = OaMakeUniquePtr<OaAdamW>(params, 0.01F);
	auto& rt     = *OaEngine::GetGlobal();

	printf("Model: ByteEmbed(%d→%d) → GRU(%d→%d, layers=1) → Linear(%d→%d)\n",
		kVocabSize, kDModel, kDModel, kHiddenDim, kHiddenDim, kVocabSize);
	printf("Params: %lld    Optimizer: AdamW(lr=0.01)\n\n",
		static_cast<long long>(model->NumParameters()));

	NlpAllPositionSampler sampler(NlpCorpus(), kBatch);
	TutorialTrainingLoop training(*opt, OaItTrainingConfig{
		.TotalSteps     = kSteps,
		.EpochSteps     = {},
		.BatchSize      = kBatch,
		.SequenceLength = kContextLen,
		.SequenceUnit   = "token",
		.SourceUnitsPerSample = kContextLen,
		.SourceUnit     = "byte",
		.TimerName      = "byte_gru_allpos_step",
		.Callbacks      = {},
	});
	printf("Training: %d steps · batch=%d · sequence=%d tokens\n", kSteps, kBatch, kContextLen);

	OaMatrix batchX;
	OaMatrix batchY;
	OaF32 initialLoss = 0.0F;

	while (not training.Loop.IsDone()) {
		sampler.NextBatch(batchX, batchY);
		opt->ZeroGrad();
		OaGradientTape tape;
		auto logits = model->Forward(batchX);
		auto loss   = OaFnLoss::CrossEntropy(logits,
			batchY.Reshape(OaMatrixShape{batchY.Size(0) * batchY.Size(1)}));
		tape.Backward(loss);
		training.Loop.Next(loss);
		if (training.Loop.Index() == 1) { initialLoss = training.Loop.LiveLoss(); }
	}
	ASSERT_TRUE(training.Loop.Finish().IsOk()) << "Finish failed";
	const OaF32 lastLoss = training.Loop.LastLoss();

	const OaF32 finalAcc = NlpAccuracyAllPositions(*model, batchX, batchY, kVocabSize);
	(void)rt;
	printf("\nEvaluation:\n");
	printf("  Random-loss baseline ln(%d) = %.4f\n",
		kVocabSize, std::log(static_cast<double>(kVocabSize)));
	printf("  Bits/byte: %.4f\n", NlpBitsPerByte(lastLoss));
	printf("  Accuracy: %.1f%%\n", finalAcc);
	printf("\nGeneration:\n  Prompt: '%s'\n  Generated: '%s'\n\n", kNlpGenerationPrompt,
		NlpGenerateGreedy(*model, kNlpGenerationPrompt, kNlpGenerationBytes, kVocabSize).c_str());

	ASSERT_GT(initialLoss, 0.0F);
	EXPECT_LT(lastLoss, initialLoss) << "Loss must decrease during training";
	EXPECT_GT(finalAcc, 50.0F)       << "All-position accuracy should exceed 50%";

	// ── Save / Load round-trip (model + optimizer) ──
	const OaString ckptPath = "/tmp/byte_gru_allpos.oam";
	ASSERT_TRUE(model->Save(ckptPath, *opt).IsOk());
	auto reloaded    = OaMakeSharedPtr<OaByteGruLM>();
	auto reloadParam = reloaded->AllParameterPtrs();
	auto reloadedOpt = OaMakeUniquePtr<OaAdamW>(reloadParam, 0.01F);
	ASSERT_TRUE(reloaded->Load(ckptPath, *reloadedOpt).IsOk());
	const OaF32 reloadedAcc = NlpAccuracyAllPositions(*reloaded, batchX, batchY, kVocabSize);
	EXPECT_NEAR(reloadedAcc, finalAcc, 0.5F)          << "Reload accuracy must match";
	EXPECT_EQ(reloadedOpt->GetStep(), opt->GetStep()) << "Optimizer step must round-trip";
}
